File: handy-ts/quiz/q13_product_of_array_except_self_medium.ts
/**
 * Q13) Product of Array Except Self
 *
 * AI-BEST:
 * Prefix and suffix products in one output array.
 * Time: O(n), Space: O(1) extra
 *
 * AI-EASY:
 * Recompute the product for each index with nested loops.
 * Time: O(n^2), Space: O(1) extra
 */

function productExceptSelfBest(nums: number[]): number[] {
  // Step 1: Build prefix products because each position needs the product of values on its left.
  const result = new Array<number>(nums.length).fill(1);
  let prefix = 1;
  for (let i = 0; i < nums.length; i++) {
    result[i] = prefix;
    prefix *= nums[i];
  }

  // Step 2: Sweep from the right with a suffix product because each position also needs the product on its right.
  let suffix = 1;
  for (let i = nums.length - 1; i >= 0; i--) {
    result[i] *= suffix;
    suffix *= nums[i];
  }

  // Step 3: Return the combined result because prefix and suffix together represent "all except self".
  return result;
}

function productExceptSelfEasy(nums: number[]): number[] {
  const result: number[] = [];

  // Step 1: For each index, walk the whole array because we are intentionally using the simplest brute-force idea.
  for (let i = 0; i < nums.length; i++) {
    let product = 1;
    for (let j = 0; j < nums.length; j++) {
      if (i === j) continue;
      product *= nums[j];
    }

    // Step 2: Store the computed product because each position is independent in this brute-force version.
    result.push(product);
  }

  // Step 3: Return all per-index products because the output array is the requested answer.
  return result;
}

function runDemo(): void {
  console.log("Q13: Product of Array Except Self");
  const cases = [
    [1, 2, 3, 4],
    [0, 1, 2, 3],
    [2, 2, 2, 2],
  ];

  for (const nums of cases) {
    console.log(`Input: nums=${JSON.stringify(nums)}`);
    console.log("  BEST:", productExceptSelfBest(nums));
    console.log("  EASY:", productExceptSelfEasy(nums));
  }
}

runDemo();

export {};

File: handy-python/quiz/q13_product_of_array_except_self_medium.py
"""
Q13) Product of Array Except Self

AI-BEST:
- Prefix products from the left and suffix products from the right.
- Time: O(n), Space: O(1) extra if output array is excluded

AI-EASY:
- Recompute the product for each index with a nested loop.
- Time: O(n^2), Space: O(1)
"""


def product_except_self_best(nums: list[int]) -> list[int]:
    # Step 1: Build prefix products into the output array.
    # Why: Each position should know the product of everything to its left.
    result = [1] * len(nums)
    prefix = 1
    for i, value in enumerate(nums):
        result[i] = prefix
        prefix *= value

    # Step 2: Multiply in suffix products while scanning from the right.
    # Why: Combining left and right contributions gives the full product except self.
    suffix = 1
    for i in range(len(nums) - 1, -1, -1):
        result[i] *= suffix
        suffix *= nums[i]

    # Step 3: Return the finished product array.
    # Why: The output now contains both prefix and suffix information for each index.
    return result


def product_except_self_easy(nums: list[int]) -> list[int]:
    # Step 1: Visit each index independently.
    # Why: This keeps the logic very direct and easy to trace.
    result: list[int] = []
    for i in range(len(nums)):
        product = 1

        # Step 2: Multiply every element except the current one.
        # Why: This matches the problem statement literally, even if it is slower.
        for j in range(len(nums)):
            if i != j:
                product *= nums[j]
        result.append(product)

    # Step 3: Return all per-index products.
    # Why: Each nested-loop pass has produced one answer slot.
    return result


def run_demo() -> None:
    print("Q13: Product of Array Except Self")
    samples = [
        [1, 2, 3, 4],
        [-1, 1, 0, -3, 3],
        [2, 3],
    ]
    for nums in samples:
        print(f"Input: nums={nums}")
        print("  BEST:", product_except_self_best(nums))
        print("  EASY:", product_except_self_easy(nums))


if __name__ == "__main__":
    run_demo()
